Add support for unsigned decimals

pjuftring opened the pull request:
This fixes #225

sunfishcode commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,19 @@ struct
+        (Rep.mul
+          (Rep.of_string (String.sub x 0 (l - 1)))
+          (Rep.of_int 10))
+        (Rep.of_string (String.sub x (l - 1) 1))
```
This code silently wraps on overflow (unless it runs out of stack space first ;-)). I think the spec should explicitly diagnose overflow.

pjuftring replied:
Didn't think of the overflow, thanks!

sunfishcode commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,19 @@ struct
+        (Rep.of_string (String.sub x (l - 1) 1))
+    in
+    try Rep.of_string x with
+      Failure _ -> unsigned_decimal x;;
```
My reading of host/lexer.mll suggests that this code code encounter a leading '-' sign, which doesn't appear to be expected in the unsigned_decimal code. Would it make sense to explicitly test for it here?

pjuftring replied:
I think I'll prohibit negative numbers in unsigned_decimal. But I'm encountering some strange stuff: lexer.mll allows '+' in front of numbers, OCaml doesn't (atleast Rep.of_string doesn't), OCaml allows '_' within numbers, lexer.mll doesn't. (This would brake my current design.) Am I missing something, or is this some not intended behavior?

pjuftring commented:
I added some overflow-tests and added the possibility to have a plus-sign in front of numbers, which was defined in host/lexer.mll, but is not allowed by the builtin of_string function.

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,35 @@ struct
+      let ten = Rep.of_int 10 in
+      let l = String.length x in
+      (* First overflow-check via String.length *)
+      if l > (if Rep.bitwidth = 32 then 10 else 20)
```
Hm, this effectively hard-codes that we only instantiate with 32 and 64 bit. It would be more modular to have a generic criterium, e.g. `bitwidth / 3`.

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,35 @@ struct
+      (* First overflow-check via String.length *)
+      if l > (if Rep.bitwidth = 32 then 10 else 20)
+        then raise (Failure "int_to_string");
+      let firstPart = Rep.of_string (String.sub x 0 (l - 1)) in
```
Style nit: Ocaml prefers snake_casing

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,35 @@ struct
+      if firstPart < Rep.zero
+        then raise (Failure "int_to_string");
+      (* Second overflow-check *)
+      if firstPart > (div_u Rep.minus_one ten)
```
Nit: redundant parens around the div_u (same below)

rossberg-chromium commented on ml-proto/spec/int.ml:
```diff
@@ -208,7 +208,35 @@ struct
+      if firstPart > (div_u Rep.minus_one ten)
+      || (firstPart = (div_u Rep.minus_one ten) && secondPart > (rem_u Rep.minus_one ten))
+        then raise (Failure "int_to_string");
+      Rep.add
```
Nit: can we put this on one line? Also, you can shorten to `Rep.(add (mul ...) ...)`

rossberg-chromium commented:
(Sorry for the delay, just rediscovered this one). I think this is good to have, but the implementation is a bit more ad-hoc than I'd hope. Wouldn't it be simpler to just reimplement the normal conversion from scratch?

pjuftring commented:
I think that would be a good alternative, this code may work, but its probably not a nice solution.
